File: go/string/30_find_substring.go
package string

// findSubstring 串联所有单词的子串
func findSubstring(s string, words []string) []int {

	// 滑动窗口思想
	// words 中的单词长度是k, 则循环k次 看s窗口中的长度k 是否在words 中存在

	if len(words) == 0 {
		return []int{}
	}

	k := len(words[0])
	n := len(words)
	sLen := len(s)

	// s 的长度 不如 words 拼接之后的长度
	if sLen < k*n {
		return []int{}
	}

	var ans = make([]int, 0)

	// k:v => word: cnt 用次来判断窗口是否满足
	wordMap := make(map[string]int)
	for _, w := range words {
		wordMap[w]++
	}

	// 循环k 次，每次步长k 相当于把所有的s 中的组合都尝试了
	for i := 0; i < k; i++ {
		// 记录窗口中已经满足单词的个数，if count == n  则记录
		count := 0
		// 记录窗口中每个单词的个数
		wordCnt := make(map[string]int)

		for l, r := i, i; r <= sLen-k; r = r + k {
			// r == sLen-k 才有可能窗口匹配到s 中的最后一个字符
			word := s[r : r+k]

			// 单词在 wordMap 中存在
			if num, ok := wordMap[word]; ok {
				// 窗口中的cnt 要超过了(数目相等)，右移 窗口左指针
				for wordCnt[word] >= num {
					wordCnt[s[l:l+k]]--
					count--
					l += k
				}

				wordCnt[word]++
				count++

			} else {
				// 单词在 wordMap 中不存在时，该窗口无效，l 到r 的下一个位置
				for l < r {
					wordCnt[s[l:l+k]]--
					count--
					l += k
				}
				// 让l 到r 的下一个位置
				l += k
			}

			if count == n {
				ans = append(ans, l)
			}
		}

	}

	return ans

}

File: rust/src/string/30_find_substring.rs
///  串联所有单词的子串
pub fn find_substring(s: String, words: Vec<String>) -> Vec<i32> {
    let mut ans = vec![];

    let n = words.len();

    if n == 0 {
        return ans;
    }

    let k = words[0].len();
    let slen = s.len();

    if slen < k * n {
        return ans;
    }

    use std::collections::HashMap;
    let mut w_map: HashMap<&str, i32> = HashMap::new();

    // 构建words 的map 用于比较
    for w in words.iter() {
        w_map.entry(w).and_modify(|v| *v += 1).or_insert(1);
    }

    // let s:Vec<char> = s.chars().collect();
    for i in 0..k {
        // 窗口中有 words 的单词数量
        let mut count = 0;

        // 记录窗口中的数量
        let mut cnt_map: HashMap<&str, i32> = HashMap::new();

        let (mut l, mut r) = (i, i);

        while r <= slen - k {
            let word = &s[r..r + k];

            // word 在 w_map 中存在
            if w_map.contains_key(word) {
                let v = w_map.get(word).unwrap();

                while let Some(&cnt) = cnt_map.get(word) {
                    // 如果窗口 中存在，并且大于cnt 不满足条件，右移 窗口左指针
                    if cnt >= *v {
                        let w = &s[l..l + k];
                        cnt_map.entry(w).and_modify(|v| *v -= 1);

                        if count > 0 {
                            count -= 1;
                        }
                        l += k;
                    } else {
                        break;
                    }
                }

                cnt_map.entry(&word).and_modify(|v| *v += 1).or_insert(1);
                count += 1;
            } else {
                // word 在 w_map 中不存在
                while l < r {
                    let w = &s[l..l + k];
                    cnt_map.entry(w).and_modify(|v| *v -= 1);

                    if count > 0 {
                        count -= 1;
                    }

                    l += k;
                }
                l += k;
            }

            r += k;

            if count == n {
                ans.push(l as i32);
            }
        }
    }

    ans
}
